apps/gimbal/src/main.cpp

```diff
@@ -1,100 +1,225 @@
+  }
+
+  bool success = true;
+
+  HWBRIDGE::CANMsgData_t msgData;
+  struct uwrt_mars_rover_can_gimbal_set_joint_pid_params_t msgStruct;
+
+  // Unpack CAN data
+  msg.getPayload(msgData);
+  if (uwrt_mars_rover_can_gimbal_set_joint_pid_params_unpack(
+          &msgStruct, msgData.raw, UWRT_MARS_ROVER_CAN_GIMBAL_SET_JOINT_PID_PARAMS_LENGTH) == 0) {
+    // Error check joint ID (only allow PID tuning of pan axis)
+    HWBRIDGE::GIMBAL_JOINT_PIDID_VALUES jointID =
+        (HWBRIDGE::GIMBAL_JOINT_PIDID_VALUES)uwrt_mars_rover_can_gimbal_set_joint_pid_params_gimbal_joint_pidid_decode(
+            msgStruct.gimbal_joint_pidid);
+
+    if (jointID != HWBRIDGE::GIMBAL_JOINT_PIDID_VALUES::GIMBAL_JOINT_PIDID_PAN) {
+      return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Extract PID params
+    float p = (float)uwrt_mars_rover_can_gimbal_set_joint_pid_params_gimbal_joint_pid_proportional_gain_decode(
+        msgStruct.gimbal_joint_pid_proportional_gain);
+    float i = (float)uwrt_mars_rover_can_gimbal_set_joint_pid_params_gimbal_joint_pid_integral_gain_decode(
+        msgStruct.gimbal_joint_pid_integral_gain);
+    float d = (float)uwrt_mars_rover_can_gimbal_set_joint_pid_params_gimbal_joint_pid_derivative_gain_decode(
+        msgStruct.gimbal_joint_pid_derivative_gain);
+    float deadzone = (float)uwrt_mars_rover_can_gimbal_set_joint_pid_params_gimbal_joint_pid_deadzone_decode(
+        msgStruct.gimbal_joint_pid_proportional_gain);
+
+    // Set PID params
+    if (auto pid = Pan::manager.getActiveController()->getPID()) {
+      if (msgStruct.gimbal_joint_pid_proportional_gain !=
+          (uint16_t)HWBRIDGE::GIMBAL_JOINT_PID_PROPORTIONAL_GAIN_VALUES::GIMBAL_JOINT_PID_PROPORTIONAL_GAIN_SNA) {
+        pid.value().get().updateProportionalGain(p);
+      }
+      if (msgStruct.gimbal_joint_pid_integral_gain !=
+          (uint16_t)HWBRIDGE::GIMBAL_JOINT_PID_INTEGRAL_GAIN_VALUES::GIMBAL_JOINT_PID_INTEGRAL_GAIN_SNA) {
+        pid.value().get().updateIntegralGain(i);
+      }
+      if (msgStruct.gimbal_joint_pid_derivative_gain !=
+          (uint16_t)HWBRIDGE::GIMBAL_JOINT_PID_DERIVATIVE_GAIN_VALUES::GIMBAL_JOINT_PID_DERIVATIVE_GAIN_SNA) {
+        pid.value().get().updateDerivativeGain(d);
+      }
+      if (msgStruct.gimbal_joint_pid_deadzone !=
+          (uint16_t)HWBRIDGE::GIMBAL_JOINT_PID_DEADZONE_VALUES::GIMBAL_JOINT_PID_DEADZONE_SNA) {
+        pid.value().get().updateDeadzone(deadzone);
+      }
+
+      // Send ACK message back
+      sendACK(HWBRIDGE::GIMBAL_ACK_VALUES::GIMBAL_ACK_GIMBAL_SET_JOINT_PID_PARAMS_ACK);
+
+    } else {
+      // PID controller doesn't exist!
+      success = false;
+    }
+  } else {
+    // Error unpacking!
+    success = false;
+  }
+
+  return success ? MBED_SUCCESS : MBED_ERROR_CODE_FAILED_OPERATION;
 }
 
-int main() {
-  printf("\r\n\r\n");
-  printf("GIMBAL APPLICATION STARTED\r\n");
-  printf("=======================\r\n");
+static mbed_error_status_t commonSwitchCANBus(CANMsg& msg) {
```

Comment:
This switch CAN bus function is common to all the targets. It might be a good idea to have this inside the CAN library